app/packages/looker/src/lookers/imavid/controller.ts

```diff
@@ -0,0 +1,245 @@
+
+    results.forEach((result, index) => {
+      if (result.status === "rejected") {
+        console.error(
+          `couldn't fetch buffer range ${this.fetchBufferManager.buffers[index]}: ${result.reason}`
+        );
+      } else {
+        this.fetchBufferManager.removeRangeAtIndex(index);
+      }
+    });
+
+    this.timeoutId = window.setTimeout(
+      this.executeFetch.bind(this),
+      BUFFERS_REFRESH_TIMEOUT_YIELD
+    );
+  }
+
+  public get currentFrameRate() {
+    return this.frameRate;
+  }
+
+  private get environment() {
+    return this.config.environment;
+  }
+
+  private get page() {
+    return this.config.page;
+  }
+
+  private get posterSampleId() {
+    return this.config.posterSample._id;
+  }
+
+  public get store() {
+    if (!ImaVidStore.has(this.posterSampleId)) {
+      ImaVidStore.set(
+        this.posterSampleId,
+        new ImaVidFrameSamples(this.storeBufferManager)
+      );
+    }
+
+    return ImaVidStore.get(this.posterSampleId);
+  }
+
+  public setFrameRate(newFrameRate: number) {
+    if (newFrameRate > 24) {
+      throw new Error("max frame rate is 24");
+    }
+
+    if (newFrameRate < 1) {
+      throw new Error("min frame rate is 1");
+    }
+
+    this.frameRate = newFrameRate;
+  }
+
+  public async fetchMore(cursor: number, count: number) {
+    const variables = this.page(cursor, count);
+
+    const fetchUid = `${this.posterSampleId}-${cursor}-${variables.count}`;
+
+    return new Promise<void>((resolve, _reject) => {
+      // do a gql query here, get samples, update store
+      this.subscription = fetchQuery<foq.paginateSamplesQuery>(
+        this.environment,
+        foq.paginateSamples,
+        variables,
+        {
+          fetchPolicy: "store-or-network",
+          networkCacheConfig: {
+            transactionId: fetchUid,
+          },
+        }
+      ).subscribe({
+        next: (data) => {
+          if (data?.samples?.edges?.length) {
+            // update store
+            for (const { cursor, node } of data.samples.edges) {
+              if (!node) {
+                continue;
+              }
+
+              if (node.__typename !== "ImageSample") {
+                throw new Error("only image samples supported");
+              }
+
+              const nodeSampleId = node.sample["_id"] as string;
+
+              this.store.samples.set(node.sample["_id"], node);
+              this.store.frameIndex.set(Number(cursor) + 1, nodeSampleId);
+              this.store.reverseFrameIndex.set(
+                nodeSampleId,
+                Number(cursor) + 1
+              );
+            }
+
+            this.storeBufferManager.addNewRange([
+              Number(data.samples.edges[0].cursor) + 1,
+              Number(data.samples.edges[data.samples.edges.length - 1].cursor) +
+                1,
+            ]);
+          }
+          resolve();
+        },
+      });
+      // todo: see if environment.retain() is applicable here,
```

Comment:
I think `retain()` is most valuable when subscribing to the relay data which could be done in the future, but unimportant right now.

Subscribing would mean the samples would update via their global store id if any other query/mutation/subscription updated the contents. Super valuable, but none of the other Lookers leverage it either at the moment